Sub-query or view cost could be wrong for cached ViewIndex

svladykin opened the pull request:
Actually the cost will be correct, but the actual cached `ViewIndex` instance, which will be used in query execution, can be outdated (use wrong indexes and join order in cached query plan) and have other cost than it was returned by `getCost`.

In this PR I return cost from cached `ViewIndex` (just taking it from the prepared query) and expire these indexes instead of `CostElement`. This should not affect performance since cost recalculation was basically the same thing as creation of new cached `ViewIndex` instance.

Also switched to `System.nanoTime()` for expiration to avoid time travel effects of `System.currentTimeMillis()`

Commit:
Sub-query or view cost could be wrong for cached ViewIndex

Commit:
CostElement class dropped

Commit:
Do not calculate eviction time for sub-queries.

Commit:
Networked test is disabled

Commit:
Merge branch 'master' of https://github.com/h2database/h2database into viewcost

Commit:
Merge branch 'master' of https://github.com/h2database/h2database into viewcost

svladykin commented:
Thomas, can you please provide feedback? Is this a valid change?

thomasmueller commented:
Getting rid of the cache sounds a bit dangerous to me, because the cache was there to improve performance. But I guess you have verified it doesn't result in a performance problem. We would (should) see that in the existing unit tests.

thomasmueller commented:
If the exiting unit tests don't show anything special, then it's OK to commit.

svladykin commented:
Actually I did not drop cost caching at all, I just moved it to existing session-local `ViewIndex` cache. 

For sub-queries it should not make any difference because initially cost cache was view-local, while that view is visible only for current query. Now we cache all the `ViewIndex` variants in session-local cache and use their cost.

For real views (which are not sub-queries) it should not affect anything as well. On the one hand I moved cost from view-local to a session-local cache (which may be considered harmful), but on another hand `ViewIndex` instance for local session must be prepared anyways if it does not exist or is outdated.

Nevertheless, I'll take a closer look at performance.

svladykin commented:
I've created simple benchmark for prepared statements with view and with sub-query. 
https://github.com/svladykin/h2benchmarks/blob/master/src/main/java/org/h2/benchmarks/PrepareView.java
In both cases my code works faster than current master:

```
master
=====

Benchmark              Mode  Cnt       Score      Error  Units
PrepareView.testView  thrpt  200  204781.716 ± 4373.829  ops/s

Benchmark                  Mode  Cnt      Score     Error  Units
PrepareView.testSubQuery  thrpt  200  24202.558 ± 313.199  ops/s

new_version
=========

Benchmark              Mode  Cnt       Score      Error  Units
PrepareView.testView  thrpt  200  212577.088 ± 2361.688  ops/s

Benchmark                  Mode  Cnt      Score     Error  Units
PrepareView.testSubQuery  thrpt  200  35432.425 ± 419.903  ops/s
```

In case of sub-query it is significantly faster because it is nomore needed to prepare the query separately for the cost cache and for index cache.

Tests look good, though there are few failures which I see in master as well. 

Namely the following tests are failing (some of them sporadically):
- TestMemoryUsage.testClob
- TestLob.testRemoveAfterDeleteAndClose
- TestLinkedTable.testLinkTable

I believe none of them are related to changes in view cost caching thus I'm merging this PR.

grandinj commented:
Nice work!
